File: akka-http-core/src/main/java/akka/http/javadsl/model/headers/HttpChallenge.java
/*
 * Copyright (C) 2009-2018 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.http.javadsl.model.headers;

import akka.http.impl.util.Util;
import akka.japi.Option;


import java.util.HashMap;
import java.util.Map;

public abstract class HttpChallenge {
    public abstract String scheme();
    public abstract String realm();

    public abstract Map<String, String> getParams();

    public static HttpChallenge create(String scheme, String realm) {
        return create(scheme, Option.option(realm));
    }

    public static HttpChallenge create(String scheme, String realm, Map<String, String> params) {
        return create(scheme, Option.option(realm), params);
    }

    public static HttpChallenge create(String scheme, Option<String> realm) {
        return akka.http.scaladsl.model.headers.HttpChallenge.apply(scheme, realm.asScala(), Util.emptyMap);
    }

    public static HttpChallenge create(String scheme, Option<String> realm, Map<String, String> params) {
        return akka.http.scaladsl.model.headers.HttpChallenge.apply(scheme, realm.asScala(), Util.convertMapToScala(params));
    }

    public static HttpChallenge createBasic(String realm) {
        Map<String, String> params = new HashMap<String, String>();
        params.put("charset", "UTF-8");
        return create("Basic", Option.option(realm), params);
    }

    public static HttpChallenge createOAuth2(String realm) {
        return create("Bearer", Option.option(realm));
    }
}

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/headers/HttpChallenge.scala
/**
 * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.http.scaladsl.model.headers

import java.util
import akka.http.javadsl.{ model ⇒ jm }
import akka.http.impl.util._
import akka.http.impl.util.JavaMapping.Implicits._

final case class HttpChallenge(scheme: String, realm: String,
                               params: Map[String, String] = Map.empty) extends jm.headers.HttpChallenge with ValueRenderable {

  def render[R <: Rendering](r: R): r.type = {
    r ~~ scheme
    if (realm != null) r ~~ " realm=" ~~#! realm
    if (params.nonEmpty) params.foreach { case (k, v) ⇒ r ~~ ',' ~~ k ~~ '=' ~~# v }
    r
  }

  /** Java API */
  def getParams: util.Map[String, String] = params.asJava
}

// FIXME: AbstractFunction3 required for bin compat. remove in Akka 10.0 and change realm in case class to option #20786
object HttpChallenge extends scala.runtime.AbstractFunction3[String, String, Map[String, String], HttpChallenge] {

  def apply(scheme: String, realm: Option[String]): HttpChallenge =
    HttpChallenge(scheme, realm.orNull, Map.empty[String, String])

  def apply(scheme: String, realm: Option[String], params: Map[String, String]): HttpChallenge =
    HttpChallenge(scheme, realm.orNull, params)
}

object HttpChallenges {

  def basic(realm: String): HttpChallenge = HttpChallenge("Basic", realm, Map("charset" → "UTF-8"))

  def oAuth2(realm: String): HttpChallenge = HttpChallenge("Bearer", realm)
}
